Generate Code, Add test, Add Examples (#3)
ballerina-platform/module-ballerinax-hubspot.crm.commerce.discounts: README.md

>To create a HubSpot Developer account, [click here](https://app.hubspot.com/signup-hubspot/developers?_ga=2.207749649.2047916093.1734412948-232493525.1734412948&step=landing_page)

### Step 1: Create HubSpot Developer Project
1. [Login](https://app.hubspot.com/login) to HubSpot developer account.

2. Create a public app by clicking on "Create app".![Building public image](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.crm.commerce.discounts/main/docs/setup/resources/build_public_app.png)

3. Click on "Create app".
![Creating App](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.crm.commerce.discounts/main/docs/setup/resources/create_app.png)

4. Under "App Info"
   - Enter Public app name.
   - Update App logo (optional).
   - Update Description (optional). 

   ![Entering App details](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.crm.commerce.discounts/main/docs/setup/resources/enter_app_details.png)

- Then move to "Auth" tab.

5. Setup the "Redirect URLs" with respective links.

>    >Example: http://localhost:9090  

Reviewer: ```suggestion
```

Author: Done.

   ![Auth Tab](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.crm.commerce.discounts/main/docs/setup/resources/auth_page.png)

Finally Create the app.

### Step 2: Get Client ID and Client secret.
Navigate to "Auth" tab.

![Client ID & Secret](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.crm.commerce.discounts/main/docs/setup/resources/client_id_secret.png)

### Step 3: Get access token and refresh token.

1. Set scopes under "Auth" tab for your app based on the [API requirements](https://developers.hubspot.com/docs/reference/api).

>    >Example: https://developers.hubspot.com/docs/reference/api/crm/commerce/discounts

Reviewer: ```suggestion
```

Author: Changed

   ![alt text](<./docs/setup/resources/exmaple_api_reference.png>)

2. Under `Auth` tab under `Sample install URL (OAuth)` section `Copy full URL`.
   >**Note:** The above copied URL is in the following format.
   ```
   https://app.hubspot.com/oauth/authorize?client_id=<client_id>&redirect_uri=<redirect_url>&scope=<scopes>
   ```

3. Choose the preferred account.

   ![alt text](<./docs/setup/resources/account_chose.png>)

   Choose account and authorize the client.

> 4. Don't panic though `This site can’t be reached` message appear. Look at the URL and find the authorization code.

Author: Let's reword this step

Author: Done

>    >**Note:** The above copied URL is in the following format.

Reviewer: ```suggestion
```

Author: Done.

   >Example: code=na1-************************S

5. Send a http request to the HubSpot.

   * Linux/MacOS

>       ```
>       curl --request POST \ 
>       --url https://api.hubapi.com/oauth/v1/token \ --header 'content-type: application/x-www-form-urlencoded' \ 
>       --data 'grant_type=authorization_code&code=<code>&redirect_uri=http://localhost:9090&client_id=<client_id>&client_secret=<client_secret>'
>       ```

Author: ```suggestion
      ```bash
      curl --request POST \ 
      --url https://api.hubapi.com/oauth/v1/token \ --header 'content-type: application/x-www-form-urlencoded' \ 
      --data 'grant_type=authorization_code&code=<code>&redirect_uri=http://localhost:9090&client_id=<client_id>&client_secret=<client_secret>'
      ```
```

Author: done

>    >Example: code=na1-**********************

Reviewer: ```suggestion
```

Author: Done.

6. Above command returns the `access token` and `refresh token`.

7. Use these tokens to authorize the client.

## Quickstart

> To use the `HubSpot CRM Commerce Discounts` connector in your Ballerina application, update the `.bal` file as follows:

Author: ```suggestion
Follow the below steps to use the `HubSpot CRM Commerce Discounts` connector in your Ballerina application
```

Author: done

### Step 1: Import the module

Import the `hubspot.crm.commerce.discounts` module and `oauth2` module.

```ballerina

> import ballerinax/hubspot.crm.commerce.discounts;

Author: Let reorder the imports.
Refer: https://learn-ballerina.github.io/best_practices/format_the_code.html?highlight=import#order-of-the-imports-in-a-bal-file

Author: done

```

### Step 2: Instantiate a new connector

1. Instantiate a `OAuth2RefreshTokenGrantConfig` with the obtained credentials and initialize the connector with it.

    ```ballerina
   configurable string clientId = ?;
   configurable string clientSecret = ?;
   configurable string refreshToken = ?;

   ConnectionConfig config = {
      auth: {
         clientId,
         clientSecret,
         refreshToken,
         credentialBearer: oauth2:POST_BODY_BEARER
      }
   };
   final Client hubSpotClient = check new (config);
   ```

2. Create a `Config.toml` file and, configure the obtained credentials obtained in the above steps as follows:

   ```toml
    clientId = <Client Id>
    clientSecret = <Client Secret>
    refreshToken = <Refresh Token>
   ```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

#### Create a New Discount

```ballerina
SimplePublicObjectInputForCreate payload = {
   associations: [],
   objectWriteTraceId: "1234",
   properties: {
      "hs_label": "test_discount",
      "hs_duration": "ONCE",
      "hs_type": "PERCENT",
      "hs_value": "40",
      "hs_sort_order": "2"
   }
};

> SimplePublicObject|error create_response = check hubspotClient->/.post(payload, {});

Author: ```suggestion
SimplePublicObject create_response = check hubspotClient->/.post(payload, {});
```

Author: Done in both README.md and tests.bal

> 

Reviewer: ```suggestion
```

Author: Done.

```

#### List all discounts

```ballerina
GetCrmV3ObjectsDiscountsQueries params = {
   'limit: 10,
   archived: false,
   properties: ["hs_label", "hs_value", "hs_type"]
};

CollectionResponseSimplePublicObjectWithAssociationsForwardPaging response = check hubspotClient->/.get({}, params);

> 

Reviewer: ```suggestion
```

Author: Resolved.
